lang/lowering/src/lookup_table/build.rs

```diff
@@ -0,0 +1,177 @@
+use ast::HashMap;
+use decls::*;
+use miette_util::ToMiette;
+use parser::cst::*;
+
+use crate::LoweringError;
+
+use super::{DeclMeta, LookupTable};
+
+pub fn build_lookup_table(module: &Module) -> Result<LookupTable, LoweringError> {
+    let mut lookup_table = LookupTable { map: HashMap::default() };
+
+    let Module { decls, .. } = module;
+
+    for decl in decls {
+        decl.build(&mut lookup_table)?;
+    }
+
+    Ok(lookup_table)
+}
+
+trait BuildLookupTable {
+    fn build(&self, lookup_table: &mut LookupTable) -> Result<(), LoweringError>;
+}
+
+impl BuildLookupTable for Decl {
+    fn build(&self, lookup_table: &mut LookupTable) -> Result<(), LoweringError> {
+        match self {
+            Decl::Data(data) => data.build(lookup_table),
+            Decl::Codata(codata) => codata.build(lookup_table),
+            Decl::Def(def) => def.build(lookup_table),
+            Decl::Codef(codef) => codef.build(lookup_table),
+            Decl::Let(tl_let) => tl_let.build(lookup_table),
+        }
+    }
+}
+
+impl BuildLookupTable for Data {
+    fn build(&self, lookup_table: &mut LookupTable) -> Result<(), LoweringError> {
+        let Data { span, name, params, ctors, .. } = self;
+        match lookup_table.map.get(name) {
+            Some(_) => {
+                return Err(LoweringError::AlreadyDefined {
```

Comment:
Ideally, this error should have two spans: one for the original definition and one for the duplicate one.

Reply:
Yes, I inlined the `add` function so that we can improve the error messages in those places. I first have to add spans to DeclMeta, though.